### PR TITLE
modify QueryScheduler to lazily acquire lanes when executing queries to avoid leaks

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
This PR fixes an issue that could occur if `druid.query.scheduler.numThreads` is configured and any exception occurs after `QueryScheduler.run` has been called to create a `Sequence`. This would result in total and/or lane specific locks being acquired, but because the sequence was not actually being evaluated, the "baggage" which typically releases these locks was not being executed. An example of how this can happen is if a group-by having filter, which wraps and transforms this sequence happens to explode while wrapping the sequence. The end result is that the locks are acquired, but never released, eventually halting the ability to execute any queries.

I've modified `QueryScheduler.run` to now use `Sequence.wrap` with a full `SequenceWrapper` implementation that acquire the locks in the `before` and releases them in the `after` to better ensure that we always release what we take, no matter what. I'm not really sure why I wasn't doing it like this before...

This PR has:

- [x] been self-reviewed.
   - [x] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/test/java/org/apache/druid/server/QueryResourceTest.java [hunk 1]
```diff
@@ -896,7 +896,9 @@ public void testTooManyQuery() throws InterruptedException
     assertAsyncResponseAndCountdownOrBlockForever(
         SIMPLE_TIMESERIES_QUERY,
         waitAllFinished,
-        response -> Assert.assertEquals(Response.Status.OK.getStatusCode(), response.getStatus())
+        response -> {
+          Assert.assertEquals(Response.Status.OK.getStatusCode(), response.getStatus());
+        }
```
Comment (imply-cheddar) [inline server/src/test/java/org/apache/druid/server/QueryResourceTest.java:901]: This looks superfluous?  I'm guessing you added some lines and then didn't clean up the `{}` or is there actually a change in here?

File: server/src/test/java/org/apache/druid/server/QueryResourceTest.java [hunk 2]
```diff
@@ -1043,19 +1045,21 @@ public <T> QueryRunner<T> getQueryRunnerForIntervals(Query<T> query, Iterable<In
...
+                    try {
+                      // pretend to be a query that is waiting on results
+                      Thread.sleep(500);
+                    }
+                    catch (InterruptedException ignored) {
+                    }
+                    // all that waiting for nothing :(
+                    return Sequences.empty();
+                  })
+              ).toList()
```
Comment (imply-cheddar) [inline server/src/test/java/org/apache/druid/server/QueryResourceTest.java:1062]: Why did you need this change in the test?  It looks like you are making the test that was doing things semi-lazily to now be explicitly eager, which feels like a behavior change?

**Review by imply-cheddar — COMMENTED** [review]
Hrm, I think I had expected to see a HavingSpec that threw an exception on `setQuery` in order to repro the behavior that we saw in the wild, but I'm not seeing that in the tests.  Just to help avoid regressions in the future, I think it would be good to have a test that throws an exception out of a Having spec and explicitly tests for what has and has not been run so far before that gets thrown  (specifically, it is thrown before any actual work has been done).

File: server/src/test/java/org/apache/druid/server/QueryResourceTest.java (hunk 2 above)
Comment (clintropolis) [reply server/src/test/java/org/apache/druid/server/QueryResourceTest.java:1062]: ah, previously the lock was grabbed eagerly, but this PR changed it to be lazy with the execution of the sequence, so these tests had to be changed to actually trigger grabbing the lock

File: server/src/test/java/org/apache/druid/server/QueryResourceTest.java (hunk 1 above)
Comment (clintropolis) [reply server/src/test/java/org/apache/druid/server/QueryResourceTest.java:901]: oops yea, i started to change something but then didn't end up doing it and forgot to roll back

**Comment by clintropolis** [comment]
>Hrm, I think I had expected to see a HavingSpec that threw an exception on `setQuery` in order to repro the behavior that we saw in the wild, but I'm not seeing that in the tests. Just to help avoid regressions in the future, I think it would be good to have a test that throws an exception out of a Having spec and explicitly tests for what has and has not been run so far before that gets thrown (specifically, it is thrown before any actual work has been done).

I guess I tested this indirectly with 
```
    Sequence<?> sequence = scheduler.run(report1, Sequences.empty());
    // making the sequence doesn't count, only running it does
    Assert.assertEquals(5, scheduler.getTotalAvailableCapacity());
```
which is asserting that the lock is no longer grabbed eagerly so anything outside of the sequence can freely explode and the right thing will happen. I'm not sure we have any test setup that has the full set of stuff and an observable `QueryScheduler` to do this exact test, will look around and see how much trouble it would be to write a test ensuring that a having filter exploding when making value matchers can't leak locks on the broker.

**Review by abhishekagarwal87 — APPROVED** [review]

**Review by abhishekagarwal87 — APPROVED** [review]
